src/sbvr-api/permissions.ts

```diff
@@ -684,6 +685,201 @@ const getAlias = (name: string) => {
+					// this relations ship means that:
+					//   mapping[0] is the local field
+					//   mapping[1] is the reference to the other resource, that joins this resource
+					//   mapping[1][0] is the name of the other resource (actor in the example)
+					//   mapping[1][1] is the name of the field on the other resource
+					//
+					// this therefore defines that the local field `actor` needs
+					// to match the `id` of the `actor` resources for the join
+
+					const possibleTargetResourceName = mapping[1][0];
+
+					// Skip this if we already shortcut this connection
+					if (possibleTargetResourceName.endsWith('$bypass')) {
+						return;
+					}
+
+					const targetResource =
+						abstractSqlModel.tables[possibleTargetResourceName];
+					const targetResourceEscaped = sqlNameToODataName(
+						targetResource?.name ?? possibleTargetResourceName,
+					);
+
+					// This is either a translated or bypassed resource we don't
+					// mess with these
+					if (targetResourceEscaped.includes('$')) {
+						return;
+					}
+
+					let foundCanAccessLink = false;
+
+					try {
+						const odata = memoizedParseOdata(`/${targetResourceEscaped}`);
+
+						const collapsedPermissionFilters = buildODataPermission(
+							permissionsLookup,
+							methodPermissions.GET,
+							vocabulary,
+							targetResourceEscaped,
+							odata,
+						);
+
+						_.set(
+							odata,
+							['tree', 'options', '$filter'],
+							collapsedPermissionFilters,
+						);
+
+						const canAccessFunction: ResourceFunction = function(
+							property: AnyObject,
+						) {
+							// remove method property so that we won't loop back here again at this point
+							delete property.method;
+
+							if (!this.defaultResource) {
+								throw new Error(`No resource selected in AST.`);
+							}
+
+							const targetResourceAST = this.NavigateResources(
+								this.defaultResource,
+								property.name,
+							);
+
+							const targetResourceName = sqlNameToODataName(
+								targetResourceAST.resource.name,
+							);
+							const currentResourceName = sqlNameToODataName(
+								this.defaultResource.name,
+							);
+
+							if (
+								currentResourceName === targetResourceEscaped &&
+								targetResourceName === escapedName
+							) {
+								foundCanAccessLink = true;
+							}
+							// return a true expression to not select the relationship, which might be virtual
+							// this should be a boolean expression, but needs to be a subquery in case it
+							// is wrapped in an `or` or `and`
+							return ['Equals', ['Boolean', true], ['Boolean', true]];
+						};
+
+						// We need execute the abstract SQL compiler to traverse
+						// through the permissions for that resource, using a
+						// special canAccess callback.
+						const odata2AbstractSQL = new OData2AbstractSQL(
+							originalAbstractSQLModel,
+							{
+								canAccess: canAccessFunction,
+							},
+						);
```

Comment:
I think creating a new OData2AbstractSQL instance is actually fairly expensive due to have to do some preprocessing on the model (for shortening aliases). This is all done once at startup right and never at runtime? if so we should be fine but we might need to look into optimizing this somehow in future to keep startup times reasonable

Reply:
This is a good point. This is executed also at runtime until the caches are hot. These should be cached per permission set. This should probably be optimized. I think this could be tricky, because we do want the scope of the current function in the canAccess function.

Reply:
Yeah I'm thinking we'll just need to be able to hotswap the canAccess function before we run the `.match` and since it's all synchronous we can know that's safe to do although it is a bit ugly